src/NServiceBus.Metrics/QueueLength/QueueLengthMarticBuilder.cs

```diff
@@ -0,0 +1,157 @@
+﻿namespace NServiceBus.Metrics.QueueLength
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Features;
+    using global::Metrics;
+    using global::Metrics.Core;
+    using Pipeline;
+    using Routing;
+
+    class QueueLengthMetricBuilder : MetricBuilder
+    {
+        const string HeaderName = "NServiceBus.Metrics.QueueLength";
+
+        const char SeparatorValue = '_';
+        static readonly string Separator = SeparatorValue.ToString();
+        static readonly char[] Separators = { SeparatorValue };
+
+        MetricsContext metricsContext;
+
+        ConcurrentDictionary<string, Tuple<Guid, CounterImplementation>> sendingCounters = new ConcurrentDictionary<string, Tuple<Guid, CounterImplementation>>();
+        ConcurrentDictionary<Guid, SequenceReporter> receivingReporters = new ConcurrentDictionary<Guid, SequenceReporter>();
+        static readonly Unit Unit = Unit.Custom("Sequence");
+
+        public override void Define(MetricsContext metricsContext)
+        {
+            this.metricsContext = metricsContext;
+        }
+
+        public override void WireUp(FeatureConfigurationContext featureConfigurationContext)
+        {
+            var pipeline = featureConfigurationContext.Pipeline;
+
+            var dispatchBehavior = new DispatchQueueLengthBehavior(this);
+            pipeline.Register(dispatchBehavior, nameof(DispatchQueueLengthBehavior));
+
+            var incomingBehavior = new IncomingQueueLengthBehavior(this);
+            pipeline.Register(incomingBehavior, nameof(IncomingQueueLengthBehavior));
+        }
+
+        Tuple<Guid, long> RegisterSend(string destination)
+        {
+            var counter = sendingCounters.GetOrAdd(destination, CreateSendCounter);
+
+            counter.Item2.Increment();
+
+            return Tuple.Create(counter.Item1, counter.Item2.Value.Count);
+        }
+
+        Tuple<Guid, CounterImplementation> CreateSendCounter(string destination)
+        {
+            var sessionId = Guid.NewGuid();
+            var counter = (CounterImplementation)metricsContext.Counter("QueueLengthSend" + Separator + destination + Separator + sessionId, Unit);
```

Comment:
On the send side we use Counter because we generate the values for a sequence.
On the receiver side we assign the value generated on the send side to a Gauge (Counters do not allow assignment, only increment).
The downside is the type mismatch (`long` vs `float`) but it is a minor inconvenience.